src/main/java/io/jenkins/plugins/httpclient/RobustHTTPClient.java

```diff
@@ -0,0 +1,258 @@
+    public interface ConnectionUser {
+        void use(CloseableHttpResponse response) throws IOException, InterruptedException;
+    }
+
+    /**
+     * Perform an HTTP network operation with appropriate timeouts and retries.
+     * @param whatConcise a short description of the operation, like {@code upload}
+     * @param whatVerbose a longer description of the operation, like {@code uploading … to …}
+     * @param connectionCreator how to establish a connection prior to getting the server’s response
+     * @param connectionUser what to do, if anything, after a successful (2xx) server response
+     * @param listener a place to print messages
+     * @throws IOException if there is an unrecoverable error; {@link AbortException} will be used where appropriate
+     * @throws InterruptedException if the transfer is interrupted
+     */
+    public void connect(String whatConcise, String whatVerbose, ConnectionCreator connectionCreator, ConnectionUser connectionUser, TaskListener listener) throws IOException, InterruptedException {
+        AtomicInteger responseCode = new AtomicInteger();
+        int attempt = 1;
+        while (true) {
+            try {
+                try {
+                    executors.submit(() -> {
+                        responseCode.set(0);
+                        try (CloseableHttpClient client = HttpClients.createSystem()) {
+                            try (CloseableHttpResponse response = connectionCreator.connect(client)) {
+                                StatusLine statusLine = response.getStatusLine();
+                                responseCode.set(statusLine != null ? statusLine.getStatusCode() : 0);
+                                if (responseCode.get() < 200 || responseCode.get() >= 300) {
+                                    String diag;
+                                    HttpEntity entity = response.getEntity();
+                                    if (entity != null) {
+                                        try (InputStream err = entity.getContent()) {
+                                            Header contentEncoding = entity.getContentEncoding();
+                                            diag = IOUtils.toString(err, contentEncoding != null ? contentEncoding.getValue() : null);
+                                        }
+                                    } else {
+                                        diag = null;
+                                    }
+                                    throw new AbortException(String.format("Failed to %s, response: %d %s, body: %s", whatVerbose, responseCode.get(), statusLine != null ? statusLine.getReasonPhrase() : "?", diag));
+                                }
+                                connectionUser.use(response);
+                            }
+                        }
+                        return null; // success
+                    }).get(timeout, TimeUnit.SECONDS);
+                } catch (TimeoutException x) {
```

Comment:
you can move it a level up and move the Exception wrapper to a method, this makes the code simpler

Reply:
Thought about doing that but did not see a way to do so without making things much more complicated.